src/workerd/api/filesystem.h

```diff
@@ -0,0 +1,333 @@
+  // they know through which path they are accessible through. Support for this
+  // is not critical so we likely won't implement this before we ship the feature
+  // but it is something we should consider doing in the future.
+  jsg::Promise<kj::Array<jsg::USVString>> resolve(
+      jsg::Lock& js, jsg::Ref<FileSystemHandle> possibleDescendant);
+
+  JSG_ITERATOR(
+      EntryIterator, entries, EntryIteratorType, IteratorState, iteratorNext<EntryIteratorType>);
+  JSG_ITERATOR(KeyIterator, keys, KeyIteratorType, IteratorState, iteratorNext<KeyIteratorType>);
+  JSG_ITERATOR(
+      ValueIterator, values, ValueIteratorType, IteratorState, iteratorNext<ValueIteratorType>);
+
+  void forEach(jsg::Lock& js,
+      jsg::Function<void(
+          jsg::USVString, jsg::Ref<FileSystemHandle>, jsg::Ref<FileSystemDirectoryHandle>)>
+          callback,
+      jsg::Optional<jsg::Value> thisArg);
+
+  JSG_RESOURCE_TYPE(FileSystemDirectoryHandle) {
+    JSG_INHERIT(FileSystemHandle);
+    JSG_METHOD(getFileHandle);
+    JSG_METHOD(getDirectoryHandle);
+    JSG_METHOD(removeEntry);
+    JSG_METHOD(resolve);
+    JSG_METHOD(entries);
+    JSG_METHOD(keys);
+    JSG_METHOD(values);
+    JSG_ITERABLE(entries);
+    JSG_METHOD(forEach);
+  }
+
+ private:
+  kj::Rc<workerd::Directory> inner;
+
+  template <typename Type>
+  static kj::Maybe<Type> iteratorNext(jsg::Lock& js, IteratorState& state) {
+    if (state.index >= state.entries.size()) {
+      return kj::none;
+    }
+
+    auto& entry = state.entries[state.index++];
+
+    if constexpr (kj::isSameType<Type, EntryIteratorType>()) {
+      return EntryType(js.accountedUSVString(entry->getName(js)), entry.addRef());
+    } else if constexpr (kj::isSameType<Type, KeyIteratorType>()) {
+      return js.accountedUSVString(entry->getName(js));
+    } else if constexpr (kj::isSameType<Type, ValueIteratorType>()) {
+      return entry.addRef();
+    } else {
+      static_assert(false, "invalid iterator type");
+    }
+  }
+};
+
+class FileSystemWritableFileStream: public WritableStream {
+  // TODO(node-fs): The spec defines FileSystemWritableFileStream objects such
+  // that any changes to the file are not actually committed until the stream
+  // is closed. We're not yet implementing this. All writes modify the source
+  // file in memory and would be immediately visible to other handles pointing
+  // to the same file. This should be updated before this feature is unflagged
+  // and made stable.
+ public:
+  struct State: public kj::Refcounted {
+    kj::Maybe<kj::Rc<workerd::File>> file;
+    size_t position = 0;
+    State(kj::Rc<workerd::File> file): file(kj::mv(file)) {}
+  };
+
+  FileSystemWritableFileStream(
+      kj::Own<WritableStreamController> controller, kj::Rc<State> sharedState);
+
+  static jsg::Ref<FileSystemWritableFileStream> constructor() = delete;
+
+  struct WriteParams {
+    kj::String type;  // one of: write, seek, truncate
+    jsg::Optional<double> size;
```

Comment:
I might be missing something but why is this double?
I think it should be uint64_t or even uint32_t

https://fs.spec.whatwg.org/#api-filesystemwritablefilestream
The SPEC seems to say unsigned long long which is not a double and double to size_t conversions in the implementation are rather ugly

Reply:
`uin64_t` maps to `bigint` in our type wrapper system. It could be `uing32_t` but the spec says these should be `unsigned long long` and not `unsigned long` which is what `uint32_t` maps to.  If we're going to map to `uint32_t` then we should change the places where we are using `size_t` to using `uint32_t` so that everything is consistent.

Reply:
Use of `uint32_t` ought to be fine throughout since that'll easily fit anything that fits without our memory limits unless we decide to support large sparse files at any point (which I wouldn't expect)

Reply:
Yeah going with uint32 makes sense to me considering our memory limitations 
Double feels wrong